### PR TITLE
A tour of KJ, and some other skeleton docs.

### PR DESCRIPTION
**Pull request by kentonv**
Most of this was written a year ago, while I was in Shanghai right when Covid was blowing up. I guess in all the commotion I forgot about the work I'd done and didn't get around to uploading it, oops... :/

### CODE REVIEW

**Comment by kentonv** [comment]
I'm really kicking myself for not having uploaded this earlier. I guess I totally forgot just how much I had written. In my memory I thought I had written only a couple paragraphs and had spent a lot more time thinking about how to structure the site, e.g. indexing the headers for reference docs. Turns out I had written a lot more than I remembered!

File: kjdoc/tour.md [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+kj::StringPtr getItem(int i) {
+  KJ_REQUIRE(i >= 0 && i < items.size()) {
+    // This is the recovery block. Recover by simply returning an empty string.
+    return "";
+  }
+  return items[i];
+}
+```
+
+When the code above is compiled with exceptions enabled, and an out-of-bounds index will result in an exception is thrown. But when compiled `-fno-exceptions`, the function will store the exception off to the side (in KJ), and then return an empty string.
```
Comment (harrishancock) [inline kjdoc/tour.md:446]: Nit: weird grammar
```suggestion
When the code above is compiled with exceptions enabled, an out-of-bounds index will result in an exception being thrown. But when compiled with `-fno-exceptions`, the function will store the exception off to the side (in KJ), and then return an empty string.
```

File: kjdoc/tour.md [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+public:
+  // ...
+
+private:
+  kj::UnwindDetector unwindDetector;
+  // ...
+};
+
+Transaction::~Transaction() noexcept(false) {
+  unwindDetector::catchExceptionsIfUnwinding([&]() {
```
Comment (harrishancock) [inline kjdoc/tour.md:505]:
```suggestion
  unwindDetector.catchExceptionsIfUnwinding([&]() {
```

**Review by a-robinson — APPROVED** [review]
This is fantastic, I'm glad you found it.

File: kjdoc/tour.md [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+
+`kj::StringPtr` represents a pointer to a `kj::String`. Similar to `kj::ArrayPtr`, `kj::StringPtr` does not point at the `kj::String` object itself, but at its backing array. Thus, moving a `kj::String` does not invalidate any `kj::StringPtr`s. This is a major difference from `std::string`! Moving an `std::string` invalidates all pointers into its backing buffer, because `std::string` inlines small strings as an optimization. This optimization may seem clever, but means that `std::string` cannot safely be used as a way to hold and transfer ownership of a text buffer. Doing so can lead to subtle, data-dependent bugs; a program might work fine until someone gives it an unusually small input, at which point it segfaults. `kj::String` foregoes this optimization for simplicity.
```
[the remark this answers is not in the document]
Comment (harrishancock) [reply kjdoc/tour.md:160]: Oh, man.

File: kjdoc/tour.md [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+}
+```
+
+The `KJ_STRINGIFY` declaration should appear either in the same namespace where the type is defined, or in the global scope. The function can return any random-access iterable sequence of `char`, such as a `kj::String`, `kj::StringPtr`, `kj::ArrayPtr<char>`, etc. As an alternative to `KJ_STRINGIFY`, you can also declare a `toString()` method on your type, with the same return type semantics.
+
+When constructing very large strings, complex strings -- for example, when writing a code generator -- consider using `kj::StringTree`, which maintains a tree of strings and only concatenates them at the very end. For example, `kj::strTree(foo, kj::strTree(bar, baz)).flatten()` only performs one concatenation, whereas `kj::str(foo, kj::str(bar, baz))` would perform a redundant intermediate concatenation.
+
+## Core Utility Types
+
+### Maybes
```
[the remark this answers is not in the document]
Comment (harrishancock) [reply kjdoc/tour.md:208]: I've hit this, too. IIRC, the behavior isn't consistent, though: `Maybe<Own<T>>` _will_ set the null state, presumably due to the optimization.

[the remark this answers is not in the document]
Comment (kentonv) [reply kjdoc/tour.md:208]: In principle you should never assume anything about a value that's been moved from. That said, maybe we should change `kj::Maybe` to consistently become null after moving away, if people tend to accidentally expect that.

File: kjdoc/tour.md [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+}
+```
+
+`kj::repeat(value, n)` returns an iterable that acts like an array of size `n` where every element is `value`. This is not often used, but can be convenient for string formatting as well as generating test data.
+
+### Casting helpers
+
+`kj::implicitCast<T>(value)` is equivalent to `static_cast<T>(value)`, but will generate a compiler error if `value` cannot be implicitly cast to `T`. For example, `static_cast` can be used for both upcasts (derived type to base type) and downcasts (base type to derived type), but `implicitCast` can only be used for the former.
+
+`kj::downcast<T>(value)` is equivalent to `static_cast<T>(value)`, except that when compiled in debug mode with RTTI available, a runtime check (`dynamic_cast`) will be performed to verify that `value` really has type `T`. Use this in cases where you are casting a base type to a derived type, and you are confident that the object is actually an instance of the derived type. The debug-mode check will hep you catch bugs.
```
Comment (harrishancock) [inline kjdoc/tour.md:98]: Typo: hep
```suggestion
`kj::downcast<T>(value)` is equivalent to `static_cast<T>(value)`, except that when compiled in debug mode with RTTI available, a runtime check (`dynamic_cast`) will be performed to verify that `value` really has type `T`. Use this in cases where you are casting a base type to a derived type, and you are confident that the object is actually an instance of the derived type. The debug-mode check will help you catch bugs.
```

**Review by harrishancock — APPROVED** [review]

**Review by bretthoerner — APPROVED** [review]
This is awesome.

**Pull request merged**